### PR TITLE
Add timeouts to client and fsms

### PR DESCRIPTION
**Pull request by russelldb**
Expose timeout to client. Use sane defaults.

No test yet as I am going to EQC the fsms when I've added R/W quora code. Passes riak_test/verify_counter_converge, tho.

### CODE REVIEW

File: src/riak_dt_update_fsm.erl [hunk 1]
```diff
@@ -108,37 +119,44 @@ prepare(timeout, SD0=#state{key=Key, from=From, mod=Mod, args=Args, req_id=ReqId
...
-%% verify it has meets consistency requirements.
+%% verify it meets consistency requirements.
 execute(timeout, SD0=#state{coord_pl_entry=CoordPLEntry,
                             preflist=PrefList,
-                            key=Key, mod=Mod,
-                            args=Args, from=From, req_id=ReqId}) ->
-    case riak_dt_vnode:update(CoordPLEntry, Mod, Key, Args) of
+                            key=Key, mod=Mod, timeout=Timeout,
+                            args=Args, req_id=ReqId}) ->
+    TRef = schedule_timeout(Timeout),
```
Comment (seancribbs) [inline src/riak_dt_update_fsm.erl]: Why wait until the `execute` state to schedule the timeout?

File: src/riak_dt_update_fsm.erl [hunk 2]
```diff
@@ -150,5 +168,24 @@ handle_sync_event(_Event, _From, _StateName, StateData) ->
...
 code_change(_OldVsn, StateName, State, _Extra) -> {ok, StateName, State}.
 
-terminate(_Reason, _SN, _SD) ->
+terminate(_Reason, _SN, #state{tref=TRef}) ->
+    cancel_timeout(TRef),
+    ok.
+
+schedule_timeout(infinity) ->
+    undefined;
+schedule_timeout(Timeout) when is_integer(Timeout) ->
```
Comment (seancribbs) [inline src/riak_dt_update_fsm.erl:177]: There's a `gen_fsm` function that might be more appropriate here than a simple `send_after`. http://www.erlang.org/doc/man/gen_fsm.html#start_timer-2

File: src/riak_dt_value_fsm.erl [hunk 3]
```diff
@@ -25,73 +25,86 @@
...
+    SD = SD0#state{ preflist = Preflist, fsm_timeout=FSMTimeout, req_timeout=ReqTimeout},
     {next_state, execute, SD, 0}.
 
-%% @doc Execute the write request and then go into waiting state to
-%% verify it has meets consistency requirements.
-execute(timeout, SD0=#state{preflist=Preflist, mod=Mod, key=Key, req_id=ReqId}) ->
+%% @doc Execute the read request and then go into waiting state for responses
+execute(timeout, SD=#state{preflist=Preflist, mod=Mod, key=Key, req_id=ReqId,
+                           req_timeout=ReqTimeout, fsm_timeout=FSMTimeout}) ->
+    TRefs = schedule_timeouts([{request_timeout, ReqTimeout}, {fsm_timeout, FSMTimeout}]),
```
Comment (seancribbs) [inline src/riak_dt_value_fsm.erl]: Same question here, why in execute and not in an earlier phase?

File: src/riak_dt_value_fsm.erl [hunk 4]
```diff
@@ -113,20 +127,26 @@ read_repair(timeout, SD=#state{mod=Mod, key=Key, replies=Replies}) ->
     do_repair(PrefList, Mod, Key, Merged),
     {stop, normal, SD}.
 
+%% @private
+handle_info(request_timeout, StateName, StateData) ->
```
Comment (seancribbs) [inline src/riak_dt_value_fsm.erl]: As I mentioned in the other file, `gen_fsm` timers won't require the use of `handle_info` but will go directly to your current state.

**Comment by seancribbs** [comment]
Other than my concerns about where timeouts are scheduled and `erlang` vs. `gen_fsm` timers, this is really good. I like especially how you refactored parts of the resources and the client module.

**Comment by seancribbs** [comment]
:+1: Merge might be tricky, but the code looks good.

**Comment by russelldb** [comment]
Merged.

**Comment by seancribbs** [comment]
:clap: :clap: :clap:

**Pull request closed without merging**